--- a/util.c.orig	2026-02-10 10:32:30.910233101 +0800
+++ b/util.c	2026-02-10 10:33:48.024018165 +0800
@@ -2580,6 +2580,38 @@
 	return ret;
 }
 
+int get_rhel_version(void)
+{
+	struct utsname name;
+	char *cp;
+	int rhel = 0;
+
+	if (uname(&name) < 0)
+		return -1;
+
+	cp = name.release;
+
+	/* 5.14.0-611.28 */
+	strtoul(cp, &cp, 10);
+	if (*cp == '.') {
+		strtoul(cp+1, &cp, 10);
+		if (*cp == '.') {
+			strtoul(cp+1, &cp, 10);
+		} else
+			return -1;
+	} else
+		return -1;
+
+	/* -611.28 */
+	if (*cp == '-') {
+		strtoul(cp+1, &cp, 10);
+		if (*cp == '.')
+			rhel = strtoul(cp+1, &cp, 10);
+	}
+
+	return rhel;
+}
+
 /* Init kernel md_mod and parameters here if needed */
 bool init_md_mod(void)
 {
@@ -2627,7 +2659,8 @@
 	 * We'll use sync mode since 6.18 rather than async mode. So in future
 	 * the kernel parameter will be removed.
 	 */
-	if (get_linux_version() >= 6018000)
+	/* kernel-5.14.0-611.28.1.el9_7 merged del sync mode */
+	if (get_rhel_version() >= 28)
 		ret = set_md_mod_parameter(MD_MOD_ASYNC_DEL_GENDISK, "N");
 
 	return ret;
